### DISCUSSION TITLE
Advice for using AutoTS with large scale grouped timeseries data

### DISCUSSION THREAD

**Discussion by bdorri** [winedarksea/AutoTS#96] [Q&A]
Firstly thank you for the wonderful package, I have enjoyed diving into it and experimenting over the last week.

Something I am still unsure about though is the best approach to take when working with a large number (1000s) of individual timeseries. For example something like the [Kaggle Walmart dataset](https://www.kaggle.com/c/walmart-sales-forecasting/data) but univariate (working on adding more series but not there yet). So far I have been looping through each timeseries separately and creating a model then doing fit(), predict(). 

I would be interested in any advice you have on this subject and whether there is a better approach with AutoTS.

Secondly, as I have univariate data I have in the past generated additional features with lags and rolling means. Is this still necessary with AutoTS or is it already happening in the background for the models that can benefit from these additional series?

**Comment by winedarksea** [comment]
Running many timeseries at once like in your dataset is exactly what AutoTS was designed for!
No `for` loops, feed them all in at once! Horizontal ensembles in particular are a flagship feature designed for this, and what they are is a method for choosing the best model for every series given, no worry about one-size-fits-all.

Some thoughts:

- set `model_interrupt=True` then you can ctrl+c (carefully) to skip over a model that is too slow
- play around at first using the `superfast` model list, it scales very well, even if it isn't the most accurate, and is good for testing params
- you can utilize the `subset=100` (or some other number) to speed things up by only testing on a subset of series (this works great if all series are similar in patterns)
- install mxnet and gluonts if you haven't already, they work very well for multivariate cases (although they can be slow to train)
- there's some more advice here: https://github.com/winedarksea/AutoTS/tree/dev#tips-for-speed-and-large-data

That said, something I am actively working on is tuning the model parameters to handle that scale well. (As in, quite literally I've got two machines running 1000's of series datasets right now to test scalability of some of the regressions).

**Reply by bdorri** [reply]
Thanks for the prompt response. 

I think I get it now :) By feeding in many series AutoTS can create an ensemble that will have different models per series based on what performs best for each particular series?

Quick follow up question: If I have data from multiple countries and I want to use `holiday_country` do I need to separate each country into it's own dataset? Or is there a way to pass in the country code in the data?

**Reply by winedarksea** [reply]
`holiday_country` is not actually used by very many models, so it doesn't matter too much. What I would suggest is, if you want to utilize multiple countries at once, is to create your own regressor of holidays (so a 0/1 flag or some such) with one column/series for each country you want the flag for, and then pass that in as part of a future_regressor.

Or you could separate into each own dataset as you suggested. The only reason to keep them together is if for the multivariate models (Regressions, VAR, GluonTS, etc) which can share information among the series.

You can borrow the `from autots.tools.holidays import holiday_flag` function or you can go to the `holidays` package directly for making a regressor.

**Reply by winedarksea** [reply]
and to answer your other question, yes. Horizontal Ensembles choose the best model per each series performance from cross validation.
